### PR TITLE
pybind/ceph_argparse: Fix error message for ceph tell command

### PR DESCRIPTION
**Pull request by neesingh-rh**
$ceph tell cephfs_1.magna023.yvyxvl client ls
error handling command target: local variable 'poolid' referenced before assignment

The error message `local variable 'poolid' referenced before assignment` looks not  descriptive. The exception was not getting raised by the usage of `ArgumentValid` which was getting bypassed and hence code asks for poolid when no type has been provided in tell commands.

Now after this fix, we are getting the proper error message as type is unknown for above example. 

Fixes: https://tracker.ceph.com/issues/59624
Signed-off-by: Neeraj Pratap Singh <neesingh@redhat.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by neesingh-rh** [comment]
After experimenting with all the classes inheriting `CephArgtype` for exception , I found this works with `JsonFormat` only. Another option that I found is to directly raise `ValueError` or 'TypeError`, it works with them too.

File: src/pybind/ceph_argparse.py
Comment (tchaikov) [inline src/pybind/ceph_argparse.py:534]: in addition to the outcome of the fix, please explain the reason *why* this change fixes the symptom you were seeing in the commit message. and a test case would be great.

Comment (tchaikov) [reply src/pybind/ceph_argparse.py:547]: per the docstring of `JsonFormat` 

> some syntactic or semantic issue with the JSON

but this is not a JSON related error. so i am inclined not using `JsonFormat` here.

File: src/pybind/ceph_argparse.py
Comment (tchaikov) [inline src/pybind/ceph_argparse.py:553]: ditto.

**Review by tchaikov — CHANGES_REQUESTED** [review]
> $ceph tell cephfs_1.magna023.yvyxvl client ls
> error handling command target: local variable 'poolid' referenced before assignment
> 
> the error message `local variable 'poolid' referenced before assignment` looks not  descriptive.
> The exception was not getting raised by the usage of `ArgumentValid` and hence code asks for poolid which is

what's the connection between `ArgumentValid` and poolid?

i still don't get why / where `poolid` is referenced

File: src/test/pybind/test_ceph_argparse.py
Comment (tchaikov) [inline src/test/pybind/test_ceph_argparse.py:352]: might want to use
```py
with self.assertRaises(SomeException):
    # do something
```

**Comment by tchaikov** [comment]
>  Error message is not descriptive for ceph tell ……command

the title of the commit message reads like a bug report. ideally, the title of the commit message should use the imperative mood. see also https://github.com/ceph/ceph/blob/master/SubmittingPatches.rst#3-describe-your-changes . please use `git commit --amend`  and `git push -f` for revising the commit message and its title.

File: src/test/pybind/test_ceph_argparse.py
Comment (batrick) [inline src/test/pybind/test_ceph_argparse.py:355]: This should be a separate test.

File: src/test/pybind/test_ceph_argparse.py
Comment (batrick) [reply src/test/pybind/test_ceph_argparse.py:352]: Agreed.

File: src/pybind/ceph_argparse.py
Comment (batrick) [reply src/pybind/ceph_argparse.py:547]: Yes, I'm left confused by this as well. Perhaps `ArgumentValid` was caught up the stack and this `JsonFormat` avoids that? This looks hacky.

**Review by batrick — CHANGES_REQUESTED** [review]

File: src/pybind/ceph_argparse.py
Comment (neesingh-rh) [reply src/pybind/ceph_argparse.py]: Yup! I went through the code again and re-debugged each flows, where I found that when `find_cmd_target()` is trying to validate the command by calling `CephName` and `CephPgid`, it is failing in the `valid()` of `CephPgid` where it is trying to access a variable `poolid` in the except clause, which isn't assigned yet due to error there. Now, the code works as expected and we are getting the proper error message for `wrong cephname type` i.e. `unknown type ....`

**Comment by neesingh-rh** [comment]
jenkins test make check

**Comment by tchaikov** [comment]
jenkins test make check

**Review by tchaikov — APPROVED** [review]
sorry for the latency. lgtm. the test failed, and the error log was removed. so i retriggered it.

File: src/pybind/ceph_argparse.py
Comment (dparmar18) [inline src/pybind/ceph_argparse.py:505]: f-strings are a better alternative here instead of format() method but looking at the entire `pybind/ceph_argparse.py` there are 70 usages of format() so merely changing these two occurrences might look odd, it could be a good idea to refactor the code but I don't think this is of high prio(?), just a heads up.

Comment (neesingh-rh) [reply src/pybind/ceph_argparse.py:505]: Yeah, I agree. I didn't change it for the same reason you mentioned. let me do this change in a separate commit/separate PR. @vshankar ?

**Comment by tchaikov** [comment]
dropping my approval
```
Traceback (most recent call last):
  File "/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/pybind/test_ceph_argparse.py", line 350, in test_tell_for_unknown_cephname_type
    with self.assertRaises(ArgumentValid):
AssertionError: ArgumentValid not raised
```

**Comment by rishabh-d-dave** [comment]
@neesingh-rh Is this PR ready for QA? Can I pick it?

**Comment by neesingh-rh** [comment]
> @neesingh-rh Is this PR ready for QA? Can I pick it?

No, there are still some issues.

**Comment by rishabh-d-dave** [comment]
> > @neesingh-rh Is this PR ready for QA? Can I pick it?
> 
> No, there are still some issues.

Okay, LMK when it's ready me.

**Comment by neesingh-rh** [comment]
vstart_runner results:
```
2024-06-19 16:31:09,288.288 INFO:__main__:Stopped test: test_ceph_tell_for_unknown_cephname_type (tasks.cephfs.test_misc.TestMisc.test_ceph_tell_for_unknown_cephname_type) in 24.5403s
2024-06-19 16:31:09,288.288 INFO:__main__:test_ceph_tell_for_unknown_cephname_type (tasks.cephfs.test_misc.TestMisc.test_ceph_tell_for_unknown_cephname_type) ... ok
2024-06-19 16:31:09,288.288 INFO:__main__:
2024-06-19 16:31:09,288.288 INFO:__main__:----------------------------------------------------------------------
2024-06-19 16:31:09,288.288 INFO:__main__:Ran 1 test in 24.541s
2024-06-19 16:31:09,288.288 INFO:__main__:
2024-06-19 16:31:09,288.288 INFO:__main__:OK
2024-06-19 16:31:09,288.288 INFO:__main__:
```

vstart_runner command execution:
```
2024-06-19 16:31:01,326.326 INFO:tasks.vstart_runner:> ./bin/ceph tell cephfs.c something
*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
2024-06-19T16:31:01.525+0530 7f192423a6c0 -1 WARNING: all dangerous and experimental features are enabled.
2024-06-19T16:31:01.536+0530 7f192423a6c0 -1 WARNING: all dangerous and experimental features are enabled.
error handling command target: unknown type cephfs
```

**Comment by neesingh-rh** [comment]
> > > @neesingh-rh Is this PR ready for QA? Can I pick it?
> > 
> > 
> > No, there are still some issues.
> 
> Okay, LMK when it's ready me.

@rishabh-d-dave It is ready for QA.

**Comment by neesingh-rh** [comment]
jenkins retest this please

**Comment by rishabh-d-dave** [comment]
This PR is under test in [https://tracker.ceph.com/issues/66753](https://tracker.ceph.com/issues/66753).

**Review by rishabh-d-dave — APPROVED** [review]
Looks good

**Review by rishabh-d-dave — APPROVED** [review]
https://tracker.ceph.com/projects/cephfs/wiki/Main#11-JUL-2024

Testing took some time because first there were issues with building testing branch, then there were too many infra failures, there were failures due to one of the PRs in the testing branch and finally last run's rerun had new failures that need to be checked.

There were no failures related to this PR, approving it.

**Comment by rishabh-d-dave** [comment]
Talked with Venky about this PR, he wants approval from core before merging.

**Comment by vshankar** [comment]
> Talked with Venky about this PR, he wants approval from core before merging.

Tagging @ljflores

**Review by tchaikov — APPROVED** [review]
lgtm

**Comment by tchaikov** [comment]
@neesingh-rh could you please rebase this change?

**Comment by neesingh-rh** [comment]
rebased!!

**Comment by neesingh-rh** [comment]
> Talked with Venky about this PR, he wants approval from core before merging.

@rishabh-d-dave We do have an approval from @tchaikov ,it is ready to merge.

**Comment by rishabh-d-dave** [comment]
> > Talked with Venky about this PR, he wants approval from core before merging.
> 
> @rishabh-d-dave We do have an approval from @tchaikov ,it is ready to merge.

Since RADOS will also be affected by this change, I suppose it's better to do RADOS QA as well? cc @vshankar @ljflores

**Comment by neesingh-rh** [comment]
> > > Talked with Venky about this PR, he wants approval from core before merging.
> > 
> > 
> > @rishabh-d-dave We do have an approval from @tchaikov ,it is ready to merge.
> 
> Since RADOS will also be affected by this change, I suppose it's better to do RADOS QA as well? cc @vshankar @ljflores

@rishabh-d-dave Can we priotize this PR for testing, as we do have BZ targeting 6.1z7 and for this blocker only is tomorrow. If we can by today, otherwise let me know. I will have reply there?

**Comment by rishabh-d-dave** [comment]
> > > > Talked with Venky about this PR, he wants approval from core before merging.
> > > 
> > > 
> > > @rishabh-d-dave We do have an approval from @tchaikov ,it is ready to merge.
> > 
> > 
> > Since RADOS will also be affected by this change, I suppose it's better to do RADOS QA as well? cc @vshankar @ljflores
> 
> @rishabh-d-dave Can we priotize this PR for testing, as we do have BZ targeting 6.1z7 and for this blocker only is tomorrow. If we can by today, otherwise let me know. I will have reply there?

From CephFS side it's already tested successfully, that's all I can do. Venky and core team are better position to tell whether to merge this or proceed for RADOS QA run.

**Comment by rishabh-d-dave** [comment]
> > > Talked with Venky about this PR, he wants approval from core before merging.
> > 
> > 
> > @rishabh-d-dave We do have an approval from @tchaikov ,it is ready to merge.
> 
> Since RADOS will also be affected by this change, I suppose it's better to do RADOS QA as well? cc @vshankar @ljflores

@vshankar @ljflores This has BZ associated with this, please check if this can be merged.

**Comment by vshankar** [comment]
> > > > Talked with Venky about this PR, he wants approval from core before merging.
> > > 
> > > 
> > > @rishabh-d-dave We do have an approval from @tchaikov ,it is ready to merge.
> > 
> > 
> > Since RADOS will also be affected by this change, I suppose it's better to do RADOS QA as well? cc @vshankar @ljflores
> 
> @vshankar @ljflores This has BZ associated with this, please check if this can be merged.

@neesingh-rh mentioned that @tchaikov has approved this, so IMO this is good to merge.

**Comment by rishabh-d-dave** [comment]
@vshankar Had a chat with Venky about this PR, we need to run other QA suites as well.

**Comment by rishabh-d-dave** [comment]
> @vshankar Had a chat with Venky about this PR, we need to run other QA suites as well.

@yuriw This PR has passed CephFS QA suits and has been approved by @tchaikov. Can you please include it in your QA runs?

**Review by vshankar — APPROVED** [review]

**Comment by vshankar** [comment]
> > @vshankar Had a chat with Venky about this PR, we need to run other QA suites as well.
> 
> @yuriw This PR has passed CephFS QA suits and has been approved by @tchaikov. Can you please include it in your QA runs?

Had a conversation with @neesingh-rh regarding this change. This started out as a different change, but ended up just as a fix in the error message. IMO, this is good to merge.

**Pull request merged**
